### PR TITLE
FED-279: Add improved prop forwarding method

### PR DESCRIPTION
**Pull request by kealjones-wk**
## Motivation
  

OverReact's prop forwarding story (fka "Consumed Props") for uiFunction (_and UiComponent2_) feels overly verbose and tedious. It also just doesn't really match the ease that comes with traditional React prop forwarding via destructuring and spread syntax in JSX, eg. `...rest`.

## Changes
  
- Added two new extension methods to UiProps
  - `getPropsToForward` for use with `addAll` (ex. `..addAll(props.getPropsToForward(exclude: {FooPropsMixin}))`)
  - `addPropsToForward` for use with `modifyProps` (ex. `..modifyProps(props.addPropsToForward(exclude: {FooPropsMixin}))`)
    - `exclude` named argument will remove the props belonging to that Set of props mixins from the forwarded props.
    - `domOnly` named argument to forward only DOM based props.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(1) Security relevant changes were detected</summary>
<li>Watched keyword <code>style=</code> in <code>web/uifunction/index.html</code> line(s) ['32'] added</li>
</details>

## Action Items
* Obtain a [security review](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); reviewer should pay special attention to insights listed above
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:194]: Should we make a form of this function that uses modifyProps and passes it to propsToUpdate so we're not making an unnecessary copy?

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:211]: Can we limit that try-catch to only the code that's relevant (i.e., just the ` final excludedProps = staticMeta.forMixins(useDefaultExclude ? {T} : exclude ?? {});` call)?

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:200]: Do we really need to support `include` for this initial implementation?

File: lib/src/component_declaration/builder_helpers.dart
Comment (kealjones-wk) [reply lib/src/component_declaration/builder_helpers.dart]: its interesting how your code example has a bunch of extra stuff lol `useDefaultExclude ? {T} : exclude ?? {}` like did you want me to add a `useDefaultExclude`? to me it seems just as easy to do `exclude: {}`

Comment (kealjones-wk) [reply lib/src/component_declaration/builder_helpers.dart]: but also, if your doing `exclude: {}` what is the point of using this method? lol

Comment (greglittlefield-wf) [reply lib/src/component_declaration/builder_helpers.dart]: That code snippet in my comment was just copy-pasted from the line that existed in this branch at the time I commented (the one with `staticMeta.forMixins`).

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:148]: I think it'd be helpful, at the expense of verbosity, to show the full component to help clarify what's going on:

> Component with a single props mixin:
> ```dart
> mixin FooPropsMixin on UiProps {
>   String foo;
> }
> 
> UiFactory<FooPropsMixin> Foo = uiFunction((props) {
>   return (Bar()
>     // Filter out props declared in FooPropsMixin 
>     // (used as the default for `exclude` since that's what `props` is statically typed as) 
>     // when forwarding to Bar.
>     ..addAll(props.getPropsToForward())
>   )();
> });
> ``` 
>
> Component with a more than one props mixin:
> ```dart
> mixin FooPropsMixin on UiProps {
>   String foo;
> }
> class FooProps = UiProps with BarProps, FooPropsMixin;
> 
> UiFactory<FooPropsMixin> Foo = uiFunction((props) {
>   return (Bar()
>     // Filter out props declared in FooPropsMixin when forwarding to Bar.
>     ..addAll(props.getPropsToForward({FooPropsMixin}))
>   )();
> });
> ```

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:143]:
```suggestion
  /// If [exclude] is not set, it defaults to using the current instance's Type.
```

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:203]: Seems like we'd want to not swallow any non-`assert` errors here
```suggestion
      rethrow;
    }
```

File: test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart:403]: Can we also verify that the `id` DOM prop is forwarded or not forwarded as expected in the other test cases?

File: web/uifunction/src/utils.dart
Comment (greglittlefield-wf) [inline web/uifunction/src/utils.dart:1]: Looks like there might be a lot of unused code here; to help keep things clean, could we remove the unused pieces?

File: web/uifunction/src/component_wrapper.dart
Comment (greglittlefield-wf) [inline web/uifunction/src/component_wrapper.dart:1]: There's a lot going on in these examples; could we add some comments explaining what they're for, for anyone who comes across them?

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:202]: Can we add test coverage for this error?

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Made some comments but generally this looks great to me!

@kealjones-wk Were you still planning on updating existing documentation on prop forwarding to recommend this new pattern? Also, would you making sure the PR description is up to date? (it still mentions `include`)

File: lib/src/component_declaration/builder_helpers.dart
Comment (kealjones-wk) [reply lib/src/component_declaration/builder_helpers.dart:202]: c75

File: web/uifunction/src/component_wrapper.dart
Comment (kealjones-wk) [reply web/uifunction/src/component_wrapper.dart:1]: can i just remove them instead because there wasnt really a reason othe than i wanted to make the visual component thing... lol

Comment (greglittlefield-wf) [reply web/uifunction/src/component_wrapper.dart:1]: Loll, up to you

File: lib/src/component_declaration/builder_helpers.dart
Comment (kealjones-wk) [reply lib/src/component_declaration/builder_helpers.dart:222]: would this not cause it to throw the error even if the above assert did not trigger? it should probably be wrapped in a like if right?

File: test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart
Comment (kealjones-wk) [reply test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart:408]: For sure, but also I honestly don't know a case where `id` would NOT be forwarded... like unless someone manually added `id` to their props mixin.

Comment (greglittlefield-wf) [reply test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart:408]: > I honestly don't know a case where id would NOT be forwarded.

That's true, assuming the implementation is correct 😉

File: doc/new_boilerplate_migration.md
Comment (kealjones-wk) [inline doc/new_boilerplate_migration.md:26]: Sorry my IDE decided to "fix weird line terminations" or something?

File: doc/new_boilerplate_migration.md
Comment (kealjones-wk) [inline doc/new_boilerplate_migration.md:784]: From here to line 894 is my actual changes

File: doc/new_boilerplate_migration.md
Comment (kealjones-wk) [reply doc/new_boilerplate_migration.md:26]: apparently i cant even undo it now....

Comment (kealjones-wk) [reply doc/new_boilerplate_migration.md]: eyyy TextEdit to the rescue!

File: doc/new_boilerplate_migration.md
Comment (kealjones-wk) [inline doc/new_boilerplate_migration.md:784]: Maybe this should go somewhere else? ugh i hate deciding where docs should go! @joebingham-wk i need your help lol

File: test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart
Comment (greglittlefield-wf) [inline test/over_react/component_declaration/builder_integration_tests/new_boilerplate/function_component_test.dart:414]: Leftover debugger (don't forget to remove the import, too)

File: lib/src/component_declaration/builder_helpers.dart
Comment (greglittlefield-wf) [inline lib/src/component_declaration/builder_helpers.dart:237]: Looks like this comment is out of date

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Not sure if this is ready for another full round of review, but I skimmed the docs changes and they look good to me, and I spotted a couple other things

**Review by greglittlefield-wf — DISMISSED** [review]
+10! 

Gonna see if anyone else on the team wants to weigh in before we merge

**Review by hunterbreathat-wk — DISMISSED** [review]

**Review by greglittlefield-wf — APPROVED** [review]
+10 refresh

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Comment by kealjones-wk** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
